op-proposer/proposer/l2_output_submitter.go

```diff
@@ -212,75 +220,250 @@ func NewL2OutputSubmitterWithSigner(
-	ctxt, cancel := context.WithTimeout(ctx, defaultDialTimeout)
-	defer cancel()
+// CreateProposalTx transforms an output response into a signed output transaction.
+// It does not send the transaction to the transaction pool.
+func (l *L2OutputSubmitter) CreateProposalTx(ctx context.Context, output *eth.OutputResponse) (*types.Transaction, error) {
+	nonce, err := l.l1Client.NonceAt(ctx, l.from, nil)
+	if err != nil {
+		l.log.Error("Failed to get nonce", "err", err, "from", l.from)
+		return nil, err
+	}
+
+	opts := &bind.TransactOpts{
+		From: l.from,
+		Signer: func(addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+			return l.signerFn(ctx, addr, tx)
+		},
+		Context: ctx,
+		Nonce:   new(big.Int).SetUint64(nonce),
+		NoSend:  true,
+	}
 
-	rpcCl, err := rpc.DialContext(ctxt, url)
+	tx, err := l.l2ooContract.ProposeL2Output(
+		opts,
+		output.OutputRoot,
+		new(big.Int).SetUint64(output.BlockRef.Number),
+		output.Status.CurrentL1.Hash,
+		new(big.Int).SetUint64(output.Status.CurrentL1.Number))
 	if err != nil {
+		l.log.Error("failed to create the ProposeL2Output transaction", "err", err)
 		return nil, err
 	}
+	return tx, nil
+}
+
+// SendTransactionExt sends a transaction through the transaction manager which handles automatic
+// price bumping.
+// It also hardcodes a timeout of 100s.
+func (l *L2OutputSubmitter) SendTransactionExt(ctx context.Context, tx *types.Transaction) error {
+	// Construct the closure that will update the txn with the current gas prices.
+	nonce := tx.Nonce()
+	updateGasPrice := func(ctx context.Context) (*types.Transaction, error) {
+		l.log.Info("proposer updating batch tx gas price", "nonce", nonce)
+		return l.UpdateGasPrice(ctx, tx)
+	}
+
+	// Wait until one of our submitted transactions confirms. If no
+	// receipt is received it's likely our gas price was too low.
+	cCtx, cancel := context.WithTimeout(ctx, 100*time.Second)
+	defer cancel()
+	receipt, err := l.txMgr.Send(cCtx, updateGasPrice, l.SendTransaction)
+	if err != nil {
+		l.log.Error("proposer unable to publish tx", "err", err)
+		return err
+	}
 
-	return sources.NewRollupClient(client.NewBaseRPCClient(rpcCl)), nil
+	// The transaction was successfully submitted
+	l.log.Info("proposer tx successfully published", "tx_hash", receipt.TxHash)
+	return nil
 }
 
-// parseAddress parses an ETH address from a hex string. This method will fail if
-// the address is not a valid hexadecimal address.
-func parseAddress(address string) (common.Address, error) {
-	if common.IsHexAddress(address) {
-		return common.HexToAddress(address), nil
+// loop is responsible for creating & submitting the next outputs
+func (l *L2OutputSubmitter) loop() {
+	defer l.wg.Done()
+
+	ctx := l.ctx
+
+	ticker := time.NewTicker(l.pollInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			cCtx, cancel := context.WithTimeout(ctx, 3*time.Minute)
+			output, shouldPropose, err := l.FetchNextOutputInfo(cCtx)
+			if err != nil {
```

Comment:
In general, I don't think we want to exit the loop on error. I've added logs for them inside the loop, though they also have logs inside the individual functions.

Reply:
Yes, sorry, we don't want to exit the loop.

OK just saw that you log every error in the called functions, like in `FetchNextOutputInfo`. I think that it's the caller's responsibility to log the error, not in the called function, which is why the error is returned back up the stack. This way you could save all the `log.Error` calls in the functions and log at the latest possible place up the stack, in `loop()`. Right now you somehow decide on how to handle errors already in the functions ("log them") but also still return them back to the caller for it to decide again.

But it's just a nit because I think it would be cleaner. No need to address it.